### PR TITLE
Fixes #36869 - Redirect /hosts to /new/hosts in index

### PR DESCRIPTION
**Pull request by parthaa**
Also removed logic in the hosts controller and application helper to rewrite the URL. Decided to go with the redirect on index approach instead since we want this redirect to happen only on get requests.

Test this
- Go to Administer-> Settings -> New Hosts Page. Set that to Yes
- Go to `http://<satellite>/hosts` and notice that it should get redirected to `http://<satellite>/new/hosts`
- Make operations like create/update/delete hosts all work ok.

### CODE REVIEW

File: app/controllers/hosts_controller.rb [hunk 1]
```diff
@@ -47,6 +47,10 @@ def index(title = nil)
     end
     respond_to do |format|
       format.html do
+        if Setting[:new_hosts_page]
+          redirect_to(new_hosts_index_page_path)
+          next
+        end
```
Comment (ekohl) [inline app/controllers/hosts_controller.rb]: If you do this, why can't it be at at the very beginning? So line 42.

**Review by ekohl — DISMISSED** [review]
I'm not a fan of this. With the host detail page we took an approach of making sure the links to it were correct, but still retained the old detail page as a fallback. This implementation means you effectively disable the whole old host page. If you go that route, why don't you make `/hosts` return the new content?

**Comment by jeremylenz** [comment]
Well, for what it's worth, I just tested this and it unbreaks host create. So :+1: in that regard.

File: app/controllers/hosts_controller.rb (hunk 1 above)
Comment (parthaa) [reply app/controllers/hosts_controller.rb]: Updated to try your approach. @jeremylenz what do you think ?

File: app/controllers/hosts_controller.rb [hunk 2]
```diff
@@ -684,7 +684,7 @@ def statuses
     render :json => statuses
   end
 
-  def hosts_path(*args)
+  def current_hosts_path(*args)
     if Setting[:new_hosts_page]
       new_hosts_index_page_path(*args)
     else
```
Comment (jeremylenz) [inline app/controllers/hosts_controller.rb]: No `super` for `current_hosts_path`, right?

Comment (parthaa) [reply app/controllers/hosts_controller.rb]: yes no super

Comment (jeremylenz) [reply app/controllers/hosts_controller.rb]: You had `super` on the next line but I couldn't comment it.

File: app/controllers/hosts_controller.rb [hunk 3]
```diff
@@ -684,11 +684,11 @@ def statuses
     render :json => statuses
   end
 
-  def hosts_path(*args)
+  def current_hosts_path(*args)
```
Comment (ekohl) [inline app/controllers/hosts_controller.rb]: Don't controllers include all helpers? So isn't this duplicating the `current_hosts_path` from `ApplicationHelper`?

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -423,11 +423,11 @@ def current_host_details_path(host)
     Setting['host_details_ui'] ? host_details_page_path(host) : host_path(host)
   end
 
-  def hosts_path(*args)
+  def current_hosts_path(*args)
```
Comment (ekohl) [inline app/helpers/application_helper.rb]: Nit: is `current` the best naming? Perhaps `preferred` is better, though it's also not great either.

**Review by jeremylenz — DISMISSED** [review]
LGTM

Host create works
also did some spot checks and links go to the correct place.

APJ (and @ekohl) :+1:

File: app/presenters/host_status_presenter.rb [hunk 5]
```diff
@@ -55,14 +55,14 @@ def global_statuses
       query = send("#{status_name}_total_query")
       return if query.empty?
 
-      Rails.application.routes.url_helpers.hosts_path(search: query)
+      Rails.application.routes.url_helpers.current_hosts_path(search: query)
```
Comment (ekohl) [inline app/presenters/host_status_presenter.rb]: Does this work?

**Review by ekohl — COMMENTED** [review]
I was late on the other PR, but one thing I really didn't like about the new host detail page was the URL change.

In your current implementation you're treating the pages as if they're equal: you can go to either one or the other. Can't we somehow implement the page in the controller to either render the old or the new version? Either based on preference or cookie.

The thing I'm mostly worried about is this scenario:
* Have the setting set to prefer the new overview
* Go to the old overview
* Use a bulk action
* Expect to be redirect back to old overview

In reality this will redirect to the new overview.

Another case I fear about is plugins. They may redirect to the hosts overview as well.

File: app/helpers/application_helper.rb (hunk 4 above)
Comment (jeremylenz) [reply app/helpers/application_helper.rb]: This is consistent with `current_host_details_path` which was already there. :shrug:

**Comment by parthaa** [comment]
> The thing I'm mostly worried about is this scenario:
> 
> * Have the setting set to prefer the new overview
> * Go to the old overview
> * Use a bulk action
> * Expect to be redirect back to old overview

You bring up an interesting idea here.

 <strike> May be we could try to be more conservative here instead of doing the `s/hosts_path/current_hosts_path`  en masse. Basically have host_controllers always return hosts_path unless the new hosts page implements those actions? . So for example right now only delete/create/change content source actions are available in New Hosts. You 'd have to go to old hosts for the rest of the actions. </strike>

Actually I just realized most of the actions on host_controller do what you were suggesting.  Something like `redirect_back_or_to hosts_path` already.  
https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L446 for example.

Which then runs 
```ruby
    def redirect_back(fallback_location:, allow_other_host: true, **args)
      referer = request.headers["Referer"]
      redirect_to_referer = referer && (allow_other_host || _url_host_allowed?(referer))
      redirect_to redirect_to_referer ? referer : fallback_location, **args
    end
```

This would mean that if the referer pointed to the old page they'd be redirected back to that and the new url would only be a fall back

**Comment by parthaa** [comment]
> The thing I'm mostly worried about is this scenario:
> 
> * Have the setting set to prefer the new overview
> * Go to the old overview
> * Use a bulk action
> * Expect to be redirect back to old overview

You bring up an interesting idea here.

 <strike> May be we could try to be more conservative here instead of doing the `s/hosts_path/current_hosts_path`  en masse. Basically have host_controllers always return hosts_path unless the new hosts page implements those actions? . So for example right now only delete/create/change content source actions are available in New Hosts. You 'd have to go to old hosts for the rest of the actions. </strike>

Actually I just realized most of the actions on host_controller do what you were suggesting.  Something like `redirect_back_or_to hosts_path` already.  
https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L446 for example.

Which then runs 
```ruby
    def redirect_back(fallback_location:, allow_other_host: true, **args)
      referer = request.headers["Referer"]
      redirect_to_referer = referer && (allow_other_host || _url_host_allowed?(referer))
      redirect_to redirect_to_referer ? referer : fallback_location, **args
    end
```

This would mean that if the referer pointed to the old page they'd be redirected back to that and the new url would only be a fall back

**Comment by parthaa** [comment]
@ekohl any thoughts on ^ ?

File: app/controllers/hosts_controller.rb (hunk 3 above)
Comment (parthaa) [reply app/controllers/hosts_controller.rb]: Updated it use `helpers.current_hosts_path`  instead

**Comment by ekohl** [comment]
That sounds like it would be less of a problem than I was afraid of, which is good.

**Comment by parthaa** [comment]
@jeremylenz can you run a quick test again on this and ack if it works

**Comment by jeremylenz** [comment]
[test unit]

**Review by jeremylenz — DISMISSED** [review]
Tested host create again and spot checked a few links, still works well for me

**Comment by jeremylenz** [comment]
[test unit]

**Comment by parthaa** [comment]
[test unit]

**Comment by parthaa** [comment]
[test unit]

**Review by jeremylenz — APPROVED** [review]

**Pull request merged**